### PR TITLE
UserPropertiesBuilder implementation.

### PR DESCRIPTION
**Pull request by SpertsyanKM**
Should be merged after user properties interfaces.

### CODE REVIEW

File: sdk/src/main/java/com/qonversion/android/sdk/UserPropertiesBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+     * Set Facebook attribution - mobile Cookie from the user's device.
+     * @param facebookAttribution Facebook attribution string
+     * @return builder instance for chain calls
+     */
+    fun setFacebookAttribution(facebookAttribution: String): UserPropertiesBuilder = apply {
+        properties[UserProperty.FacebookAttribution.code] = facebookAttribution
+    }
+
+    /**
+     * Set user property with custom key different from defined ones.
```
Comment (Maria-Bordunova) [inline sdk/src/main/java/com/qonversion/android/sdk/UserPropertiesBuilder.kt:80]:
```suggestion
     * Set a user property with a custom key different from defined ones.
```

File: sdk/src/main/java/com/qonversion/android/sdk/UserPropertiesBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+     * @param key nonempty key for user property consisting of letters A-Za-z, numbers, and symbols _.:-
+     * @param value nonempty value for the given property
+     */
+    fun setCustomUserProperty(key: String, value: String): UserPropertiesBuilder = apply {
+        properties[key] = value
+    }
+
+    /**
+     * Build final properties map with all the properties provided to builder before.
```
Comment (Maria-Bordunova) [inline sdk/src/main/java/com/qonversion/android/sdk/UserPropertiesBuilder.kt:89]:
```suggestion
     * Build final properties map with all the properties provided to the builder before.
```

File: sdk/src/test/java/com/qonversion/android/sdk/UserPropertiesBuilderTest.kt [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    @Test
+    fun `set name`() {
+        // given
+        val name = "test name"
+
+        // when
+        builder.setName(name)
+
+        // then
+        assertThat(builder.properties[UserProperty.Name.code]).isEqualTo(name)
```
Comment (Maria-Bordunova) [inline sdk/src/test/java/com/qonversion/android/sdk/UserPropertiesBuilderTest.kt:27]: Should we use "_q_name" or UserProperty.Name.code for the key?

File: sdk/src/main/java/com/qonversion/android/sdk/Qonversion.kt [hunk 4]
```diff
@@ -110,10 +110,11 @@ interface Qonversion {
...
      * as well as to provide it to third-party platforms.
-     * This method consumes both defined and custom user properties. For defined properties
-     * use [UserProperty.code] as the key. In order to pass single property
-     * consider using [setCustomUserProperty] method for the custom keys and [setUserProperties]
-     * for defined ones.
+     * This method consumes both defined and custom user properties. Consider using
+     * [UserPropertiesBuilder] to prepare a properties map. You are also able to create it
+     * on your own using [UserProperty.code] as the key for Qonversion defined properties.
+     * In order to pass single property consider using [setCustomUserProperty] method for
+     * the custom keys and [setUserProperties] for defined ones.
```
Comment (suriksarkisyan) [inline sdk/src/main/java/com/qonversion/android/sdk/Qonversion.kt:117]: I'm not sure, but I think you mean `and [setUserProperty] for defined ones.` not `setUserProperties`

File: sdk/src/main/java/com/qonversion/android/sdk/UserPropertiesBuilder.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
+package com.qonversion.android.sdk
+
+import com.qonversion.android.sdk.dto.UserProperty
+
+/**
+ * This builder class can be used to generate a map of user properties
+ * which can be then provided to [Qonversion.setUserProperties].
+ * It consumes both Qonversion defined and custom properties.
+ */
```
Comment (suriksarkisyan) [inline sdk/src/main/java/com/qonversion/android/sdk/UserPropertiesBuilder.kt:9]: I think we need to clarify that the same key of the property will overwrite the previous property value.
For example, if you call `setName` twice, only the last value is set to the dictionary.
And make attention that possible to call `setCustomUserProperty` as many times as you need. Just use different keys.

File: sdk/src/main/java/com/qonversion/android/sdk/UserPropertiesBuilder.kt [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+     * @param name name to set to the current user
+     * @return builder instance for chain calls
+     */
+    fun setName(name: String): UserPropertiesBuilder = apply {
+        properties[UserProperty.Name.code] = name
+    }
+
+    /**
+     * Set custom user id. It can be an identifier used on your backend
+     * to tie the current Qonversion user with your one.
```
[the remark this answers is not in the document]
Comment (suriksarkisyan) [reply sdk/src/main/java/com/qonversion/android/sdk/UserPropertiesBuilder.kt:25]: Let's clarify again and say that a custom user id must be a unique identifier.

File: sdk/src/test/java/com/qonversion/android/sdk/UserPropertiesBuilderTest.kt (hunk 3 above)
Comment (SpertsyanKM) [reply sdk/src/test/java/com/qonversion/android/sdk/UserPropertiesBuilderTest.kt:27]: I guess no, as we don't test concrete enum value code

**Review by suriksarkisyan — APPROVED** [review]

**Review by Maria-Bordunova — APPROVED** [review]

**Pull request merged**